### PR TITLE
Add csr struct to processor state

### PR DESCRIPTION
**Pull request by Johnabell**
This PR adds initial support for the control and status register instructions.

TODO:
- [x] Add CSR structs and triats
- [x] Implement CSRRW
- [x] Implement CSRRS
- [x] Implement CSRRC
- [x] Implement CSRRWI
- [x] Implement CSRRSI
- [x] Implement CSRRCI
- [x] Implement psuedo instructions associated with CSR

### CODE REVIEW

File: src/csr.rs [hunk 1]
```diff
@@ -0,0 +1,166 @@
...
+                self.registers[index as usize].fetch_and(!value, SeqCst)
+            }
+        }
+    };
+}
+
+#[cfg(test)]
+impl PartialEq for CSR32 {
+    fn eq(&self, other: &Self) -> bool {
+        for (index, value) in self.registers.iter().enumerate() {
```
Comment (andybalaam) [inline src/csr.rs]: Could zip here

Comment (andybalaam) [reply src/csr.rs]: Also, should handle different lengths for completeness

File: src/csr.rs [hunk 2]
```diff
@@ -0,0 +1,166 @@
...
+    }
+}
+
+implement_csr!(CSR32, i32);
+implement_csr!(CSR64, i64);
+
+#[cfg(test)]
+mod test {
+    use super::*;
+
```
Comment (andybalaam) [inline src/csr.rs:134]: Need some tests for read

Comment (andybalaam) [reply src/csr.rs:134]: Also read_write

File: src/csr.rs [hunk 3]
```diff
@@ -0,0 +1,166 @@
...
+        let csr_64 = CSR32::default();
+
+        assert_eq!(csr_32.set_bits(42, -1), 0);
+        assert_eq!(csr_64.set_bits(42, -1), 0);
+        assert_eq!(csr_32.set_bits(42, 0), -1);
+        assert_eq!(csr_64.set_bits(42, 0), -1);
+        assert_eq!(csr_32.set_bits(42, -1), -1);
+        assert_eq!(csr_64.set_bits(42, -1), -1);
+        assert_eq!(csr_32.set_bits(42, 0), -1);
+        assert_eq!(csr_64.set_bits(42, 0), -1);
```
Comment (andybalaam) [inline src/csr.rs]: Good to check that a logical or is being done

File: src/csr.rs [hunk 4]
```diff
@@ -0,0 +1,166 @@
...
+    fn clear_bits() {
+        let csr_32 = CSR32::default();
+        let csr_64 = CSR32::default();
+
+        assert_eq!(csr_32.set_bits(42, -1), 0);
+        assert_eq!(csr_64.set_bits(42, -1), 0);
+        assert_eq!(csr_32.clear_bits(42, -1), -1);
+        assert_eq!(csr_64.clear_bits(42, -1), -1);
+        assert_eq!(csr_32.set_bits(42, 0), 0);
+        assert_eq!(csr_64.set_bits(42, 0), 0);
```
Comment (andybalaam) [inline src/csr.rs]: Good to check for logical and

File: src/csr.rs [hunk 5]
```diff
@@ -0,0 +1,166 @@
...
+///   the number of instructions retired by this hart from some arbitrary start
+///   point in the past.
+///
+/// Additionally some registers are read only.
+#[derive(Debug)]
+pub struct CSR64 {
+    registers: Box<[AtomicI64]>,
+}
+
+macro_rules! implement_csr {
```
Comment (andybalaam) [inline src/csr.rs:78]: Document that `Self::registers` needs to be a `Box<AtomicX>`

File: src/instructions/csr.rs [hunk 6]
```diff
@@ -0,0 +1,33 @@
+use std::ops::Deref;
+
```
Comment (andybalaam) [inline src/instructions/csr.rs:2]: Document this: it's for decoding the binary CSR instruction

File: src/csr.rs [hunk 7]
```diff
@@ -0,0 +1,166 @@
+use std::sync::atomic::{AtomicI32, AtomicI64, Ordering::SeqCst};
+
+/// According to the RISC-V specification, the number of control status registers.
+const CSR_SIZE: usize = 4096;
+
+/// The control status registers.
+pub trait CSR {
```
Comment (andybalaam) [inline src/csr.rs]: Rename to `ControlStatusRegisters`

File: src/instructions/pseudoinstructions.rs [hunk 8]
```diff
@@ -211,6 +211,123 @@ impl Instruction {
...
+        PseudoinstructionMappingIter::One(Instruction::CSRRW {
+            rd,
+            rs1: Register::ZERO,
+            csr,
+        })
+    }
+
+    /// # Write CSR
+    ///
+    /// Write CSR, no read side affects should be caused by this instruction.
```
Comment (andybalaam) [inline src/instructions/pseudoinstructions.rs]: effects

File: src/instructions/pseudoinstructions.rs [hunk 9]
```diff
@@ -211,6 +211,123 @@ impl Instruction {
...
+        PseudoinstructionMappingIter::One(Instruction::CSRRW {
+            rd: Register::ZERO,
+            rs1,
+            csr,
+        })
+    }
+
+    /// # Set bits in CSR
+    ///
+    /// Sets the bits in CSR, no read side affects should be caused by this
```
Comment (andybalaam) [inline src/instructions/pseudoinstructions.rs]: effects

File: src/instructions/pseudoinstructions.rs [hunk 10]
```diff
@@ -211,6 +211,123 @@ impl Instruction {
...
+        PseudoinstructionMappingIter::One(Instruction::CSRRS {
+            rd: Register::ZERO,
+            rs1,
+            csr,
+        })
+    }
+
+    /// # Clear bits in CSR
+    ///
+    /// Clear bits in CSR, no read side affects should be caused by this
```
Comment (andybalaam) [inline src/instructions/pseudoinstructions.rs]: effects

File: src/registers.rs [hunk 11]
```diff
@@ -356,4 +394,53 @@ mod test {
...
+        let register_zero = ZeroRegister::<i32>::default();
+        assert_eq!(*register_zero, 0);
+    }
+
+    #[test]
+    fn register_zero_deref_mut() {
+        let mut register_zero = ZeroRegister::<i32>::default();
+        *register_zero = 23;
+        assert_eq!(*register_zero, 0);
+    }
```
Comment (andybalaam) [inline src/registers.rs:419]: Test for deref_mut again seeing the old value

**Review by andybalaam — APPROVED** [review]

File: src/csr.rs (hunk 5 above)
Comment (Johnabell) [reply src/csr.rs:78]: Done

File: src/csr.rs (hunk 1 above)
Comment (Johnabell) [reply src/csr.rs]: Done

File: src/csr.rs (hunk 2 above)
Comment (Johnabell) [reply src/csr.rs:134]: Done

File: src/csr.rs (hunk 3 above)
Comment (Johnabell) [reply src/csr.rs]: Covered elsewhere

File: src/csr.rs (hunk 4 above)
Comment (Johnabell) [reply src/csr.rs]: Covered elsewhere

File: src/instructions/csr.rs (hunk 6 above)
Comment (Johnabell) [reply src/instructions/csr.rs:2]: We might have a look at making some of this code reusable in a future PR

File: src/registers.rs (hunk 11 above)
Comment (Johnabell) [reply src/registers.rs:419]: Done

**Pull request merged**